lib/services/local/dynamic_access.go

```diff
@@ -339,15 +345,48 @@ func (s *DynamicAccessService) getAccessRequestPluginData(ctx context.Context, f
 
 // UpdatePluginData updates a per-resource PluginData entry.
 func (s *DynamicAccessService) UpdatePluginData(ctx context.Context, params types.PluginDataUpdateParams) error {
-	switch params.Kind {
-	case types.KindAccessRequest:
-		return trace.Wrap(s.updateAccessRequestPluginData(ctx, params))
-	default:
+	if !isAllowedPluginDataKind(params.Kind) {
 		return trace.BadParameter("unsupported resource kind %q", params.Kind)
 	}
+
+	return trace.Wrap(s.updatePluginData(ctx, params))
+}
+
+// getResourceExpiry returns a resource expiration time
+func (s *DynamicAccessService) getResourceExpiry(ctx context.Context, resource string, kind string) (time.Time, error) {
```

Comment:
Is there some sort of interfaced-based approach we could take to this? E.g. a `ExpiryGetter` interface that is implemented by all resource types that you need to get an expiry for?

**Edit:** 
Actually, that would
1. be overkill for now - the switch does the job fine, and
2. probably mean you'd have to pull the arbitrary resource out of the back end,  un-marshall it, query for the correct interface and only _then_ pull the expiry out. If the expiry just uses the backend record's expiry value (as it does in a full 50% of the covered cases), that would be a stupid amount of overwork. 

Leaving the comment here for the sake of discussion, but consider it resolved.

Reply:
Yes, there actually is. I did one more pass over the existing architecture. I think, I now understand it better. I have re-implemented `PluginData` from scratch as a separate service. Now, there is isolated `PluginData` service which is responsible for storage only, and there are `Server` methods which are responsible for high-level checks. Thank you for pointing to it.